api/docker-entrypoint.sh

```diff
@@ -0,0 +1,177 @@
+#!/bin/bash
+
+info()
+{
+    echo "Info: $*"
+}
+
+warn()
+{
+    echo "Warning: $*"
+}
+
+fatal()
+{
+    echo "Fatal: $*"
+}
+
+get_couchdb_url()
+{
+  echo "http://$COUCHDB_USER:$COUCHDB_PASSWORD@$COUCHDB_SERVICE_NAME:5985"
+}
+
+wait_for_couchdb()
+{
+  local n=60
+
+  while [ "$n" -gt 0 ]; do
+    curl "`get_couchdb_url`" &>/dev/null
+    [ "$?" -eq 0 ] && return 0
+    sleep 3
+    n=$[$n-1]
+  done
+
+  return 1
+}
+
+is_setup_needed()
+{
+  ! is_existing_user 'medic-api'
+}
+
+is_existing_user()
+{
+  local user="$1"
+  local url="`get_couchdb_url`"
+  local id="org.couchdb.user:$user"
+  shift 1
+
+  local userdoc=$(curl -X GET $url/_users/$id | jq '._id?')
+
+  if [ "$userdoc" = \"org.couchdb.user:$user\" ]; then
+    if [ ! -f /srv/storage/$user/passwd/$user ]; then
+      create_couchdb_admin "$user"
+      return 1
+    fi
+    return 0
+  else
+    return 1
+  fi
+}
+
+create_couchdb_put()
+{
+  local should_auth="$1"
+  local payload="$2"
+  local payload_is_file="$3"
+
+  local cmd='cat'
+
+  if [ -z "$payload_is_file" ]; then
+    cmd='echo'
+  fi
+
+  if [ "$#" -gt 1 ]; then
+    echo -n 'data = "\"' &&
+    "$cmd" "$payload" &&
+    echo '\""'
+  fi
+}
+
+create_couchdb_admin()
```

Comment:
Do you think that it would be better to do this work in a couchdb ini file and copy that in to the specific couch container? That would allow the couchdb container to come up pre-configured vs having to do it in each entrypoint script.

The section about Using a persistent CouchDB configuration file [here](https://hub.docker.com/_/couchdb) details the setup.

Reply:
That's how I had it written it previously, but then we would need to:
1) share that password as a volume between containers, so API can read its service password
2) Or offload that to the user/SRE and have them generate a password and export it as a variable (and save that password as a variable during all restarts) for each service. I was worried about this being an unnecessary burden on self-host small teams.

**1** would really impact benefits we receive in k8s. We would have to deploy all containers sharing that volume. I think **2** is do-able, but I was worried about the extra step during all restart scenarios in both deployments (self-hosted and medic-hosted). And we get a bonus in the current entrypoint that the service users regenerate a new password on restarts.

I overcame the security fears in that we will use internal DNS (k8s), so nothing travels across public http. Self-host will have the same thing (localhost).

If there's a method that covers those scenarios but uses .ini, I'd prefer that too!